File: php/core/Context.php
<?php
declare(strict_types=1);

// YugiLimitRegulation SDK context

require_once __DIR__ . '/Control.php';
require_once __DIR__ . '/Operation.php';
require_once __DIR__ . '/Spec.php';
require_once __DIR__ . '/Result.php';
require_once __DIR__ . '/Response.php';
require_once __DIR__ . '/Error.php';
require_once __DIR__ . '/Helpers.php';

class YugiLimitRegulationContext
{
    public string $id;
    public array $out;
    public mixed $client;
    public ?YugiLimitRegulationUtility $utility;
    public YugiLimitRegulationControl $ctrl;
    public array $meta;
    public ?array $config;
    public ?array $entopts;
    public ?array $options;
    public mixed $entity;
    public ?array $shared;
    public array $opmap;
    public array $data;
    public array $reqdata;
    public array $match;
    public array $reqmatch;
    public ?array $point;
    public ?YugiLimitRegulationSpec $spec;
    public ?YugiLimitRegulationResult $result;
    public ?YugiLimitRegulationResponse $response;
    public YugiLimitRegulationOperation $op;

    public function __construct(array $ctxmap = [], ?self $basectx = null)
    {
        $this->id = 'C' . random_int(10000000, 99999999);
        $this->out = [];

        $this->client = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'client') ?? ($basectx ? $basectx->client : null);
        $this->utility = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'utility') ?? ($basectx ? $basectx->utility : null);

        $this->ctrl = new YugiLimitRegulationControl();
        $ctrl_raw = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'ctrl');
        if (is_array($ctrl_raw)) {
            if (array_key_exists('throw', $ctrl_raw)) {
                $this->ctrl->throw_err = $ctrl_raw['throw'];
            }
            if (isset($ctrl_raw['explain']) && is_array($ctrl_raw['explain'])) {
                $this->ctrl->explain = $ctrl_raw['explain'];
            }
        } elseif ($basectx !== null && $basectx->ctrl !== null) {
            $this->ctrl = $basectx->ctrl;
        }

        $m = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'meta');
        $this->meta = is_array($m) ? $m : ($basectx ? $basectx->meta ?? [] : []);

        $cfg = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'config');
        $this->config = is_array($cfg) ? $cfg : ($basectx ? $basectx->config : null);

        $eo = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'entopts');
        $this->entopts = is_array($eo) ? $eo : ($basectx ? $basectx->entopts : null);

        $o = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'options');
        $this->options = is_array($o) ? $o : ($basectx ? $basectx->options : null);

        $e = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'entity');
        $this->entity = $e ?? ($basectx ? $basectx->entity : null);

        $s = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'shared');
        $this->shared = is_array($s) ? $s : ($basectx ? $basectx->shared : null);

        $om = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'opmap');
        $this->opmap = is_array($om) ? $om : ($basectx ? $basectx->opmap ?? [] : []);

        $this->data = YugiLimitRegulationHelpers::to_map(YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'data')) ?? [];
        $this->reqdata = YugiLimitRegulationHelpers::to_map(YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'reqdata')) ?? [];
        $this->match = YugiLimitRegulationHelpers::to_map(YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'match')) ?? [];
        $this->reqmatch = YugiLimitRegulationHelpers::to_map(YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'reqmatch')) ?? [];

        $pt = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'point');
        $this->point = is_array($pt) ? $pt : ($basectx ? $basectx->point : null);

        $sp = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'spec');
        $this->spec = ($sp instanceof YugiLimitRegulationSpec) ? $sp : ($basectx ? $basectx->spec : null);

        $r = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'result');
        $this->result = ($r instanceof YugiLimitRegulationResult) ? $r : ($basectx ? $basectx->result : null);

        $rp = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'response');
        $this->response = ($rp instanceof YugiLimitRegulationResponse) ? $rp : ($basectx ? $basectx->response : null);

        $opname = YugiLimitRegulationHelpers::get_ctx_prop($ctxmap, 'opname') ?? '';
        $this->op = $this->resolve_op($opname);
    }

    public function resolve_op(string $opname): YugiLimitRegulationOperation
    {
        if (isset($this->opmap[$opname])) {
            return $this->opmap[$opname];
        }
        if ($opname === '') {
            return new YugiLimitRegulationOperation([]);
        }

        $entname = (is_object($this->entity) && method_exists($this->entity, 'get_name'))
            ? $this->entity->get_name()
            : '_';
        $opcfg = \Voxgig\Struct\Struct::getpath($this->config, "entity.{$entname}.op.{$opname}");

        $input = ($opname === 'update' || $opname === 'create') ? 'data' : 'match';

        $points = [];
        if (is_array($opcfg)) {
            $t = \Voxgig\Struct\Struct::getprop($opcfg, 'points');
            if (is_array($t)) {
                $points = $t;
            }
        }

        $op = new YugiLimitRegulationOperation([
            'entity' => $entname,
            'name' => $opname,
            'input' => $input,
            'points' => $points,
        ]);
        $this->opmap[$opname] = $op;
        return $op;
    }

    public function make_error(string $code, string $msg): YugiLimitRegulationError
    {
        return new YugiLimitRegulationError($code, $msg, $this);
    }
}

File: .sdk/tm/rb/core/context.rb
# YugiLimitRegulation SDK context

require_relative '../utility/struct/voxgig_struct'
require_relative 'control'
require_relative 'operation'
require_relative 'spec'
require_relative 'result'
require_relative 'response'
require_relative 'error'
require_relative 'helpers'

class YugiLimitRegulationContext
  attr_accessor :id, :out, :client, :utility, :ctrl, :meta, :config,
                :entopts, :options, :entity, :shared, :opmap,
                :data, :reqdata, :match, :reqmatch, :point,
                :spec, :result, :response, :op

  def initialize(ctxmap = {}, basectx = nil)
    ctxmap ||= {}
    @id = "C#{rand(10000000..99999999)}"
    @out = {}

    @client = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "client") || basectx&.client
    @utility = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "utility") || basectx&.utility

    @ctrl = YugiLimitRegulationControl.new
    ctrl_raw = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "ctrl")
    if ctrl_raw.is_a?(Hash)
      @ctrl.throw_err = ctrl_raw["throw"] if ctrl_raw.key?("throw")
      @ctrl.explain = ctrl_raw["explain"] if ctrl_raw["explain"].is_a?(Hash)
    elsif basectx&.ctrl
      @ctrl = basectx.ctrl
    end

    m = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "meta")
    @meta = m.is_a?(Hash) ? m : (basectx&.meta || {})

    cfg = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "config")
    @config = cfg.is_a?(Hash) ? cfg : basectx&.config

    eo = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "entopts")
    @entopts = eo.is_a?(Hash) ? eo : basectx&.entopts

    o = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "options")
    @options = o.is_a?(Hash) ? o : basectx&.options

    e = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "entity")
    @entity = e || basectx&.entity

    s = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "shared")
    @shared = s.is_a?(Hash) ? s : basectx&.shared

    om = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "opmap")
    @opmap = om.is_a?(Hash) ? om : (basectx&.opmap || {})

    @data = YugiLimitRegulationHelpers.to_map(YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "data")) || {}
    @reqdata = YugiLimitRegulationHelpers.to_map(YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "reqdata")) || {}
    @match = YugiLimitRegulationHelpers.to_map(YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "match")) || {}
    @reqmatch = YugiLimitRegulationHelpers.to_map(YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "reqmatch")) || {}

    pt = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "point")
    @point = pt.is_a?(Hash) ? pt : basectx&.point

    sp = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "spec")
    @spec = sp.is_a?(YugiLimitRegulationSpec) ? sp : basectx&.spec

    r = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "result")
    @result = r.is_a?(YugiLimitRegulationResult) ? r : basectx&.result

    rp = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "response")
    @response = rp.is_a?(YugiLimitRegulationResponse) ? rp : basectx&.response

    opname = YugiLimitRegulationHelpers.get_ctx_prop(ctxmap, "opname") || ""
    @op = resolve_op(opname)
  end

  def resolve_op(opname)
    return @opmap[opname] if @opmap[opname]
    return YugiLimitRegulationOperation.new({}) if opname.empty?

    entname = @entity&.respond_to?(:get_name) ? @entity.get_name : "_"
    opcfg = VoxgigStruct.getpath(@config, "entity.#{entname}.op.#{opname}")

    input = (opname == "update" || opname == "create") ? "data" : "match"

    points = []
    if opcfg.is_a?(Hash)
      t = VoxgigStruct.getprop(opcfg, "points")
      points = t if t.is_a?(Array)
    end

    op = YugiLimitRegulationOperation.new({
      "entity" => entname,
      "name" => opname,
      "input" => input,
      "points" => points,
    })
    @opmap[opname] = op
    op
  end

  def make_error(code, msg)
    YugiLimitRegulationError.new(code, msg, self)
  end
end
